### PR TITLE
[Feature] Add BezierAlign CUDA op

### PR DESCRIPTION
**Pull request by Harold-lkk**
## Motivation

Bezier Align for ABCNet

### CODE REVIEW

**Comment by grimoire** [comment]
Please update docs in `docs/en/understand_mmcv/ops.md` and and `docs/zh_cn/understand_mmcv/ops.md`, Add unit test in `tests/test_ops/`

File: mmcv/ops/csrc/pytorch/cuda/cudabind.cpp [hunk 1]
```diff
@@ -1867,3 +1867,46 @@ REGISTER_DEVICE_IMPL(prroi_pool_forward_impl, CUDA, prroi_pool_forward_cuda);
...
+                                       float spatial_scale, int sampling_ratio,
+                                        bool aligned);
+
+void BezierAlignBackwardCUDAKernelLauncher(Tensor grad_output, Tensor rois,
+                                        Tensor grad_input, int aligned_height,
+                                        int aligned_width, float spatial_scale,
+                                        int sampling_ratio,
+                                        bool aligned);
+
+void bezier_align_forward_cuda(Tensor input, Tensor rois, Tensor output,
```
Comment (grimoire) [inline mmcv/ops/csrc/pytorch/cuda/cudabind.cpp:1883]: `BezierAlignForwardCUDAKernelLauncher` and `bezier_align_forward_cuda` share the same arguments. We do not need to wrap it again. Just register it with:
`REGISTER_DEVICE_IMPL(bezier_align_forward_impl, CUDA, BezierAlignForwardCUDAKernelLauncher);`

File: mmcv/ops/bezier_align.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        aligned: bool = True,
+    ):
+        super().__init__()
+
+        self.output_size = _pair(output_size)
+        self.spatial_scale = float(spatial_scale)
+        self.sampling_ratio = int(sampling_ratio)
+        self.aligned = aligned
+
+    def forward(self, input, beziers):
```
Comment (grimoire) [inline mmcv/ops/bezier_align.py:119]: Add docstring here.

File: mmcv/ops/bezier_align.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+            spatial_scale=ctx.spatial_scale,
+            sampling_ratio=ctx.sampling_ratio,
+            aligned=ctx.aligned)
+
+        ctx.save_for_backward(beziers)
+        return output
+
+    @staticmethod
+    @once_differentiable
+    def backward(ctx, grad_output):
```
Comment (grimoire) [inline mmcv/ops/bezier_align.py:51]: Please add type hint.

File: mmcv/ops/csrc/pytorch/cpu/bezier_align.cpp [hunk 4]
```diff
@@ -0,0 +1,567 @@
...
+                               int aligned_height, int aligned_width,
+                               float spatial_scale, int sampling_ratio,
+                               bool aligned);
+
+void bezier_align_backward_impl(Tensor grad_output, Tensor rois, Tensor grad_input,
+                                int aligned_height, int aligned_width,
+                                float spatial_scale, int sampling_ratio,
+                                bool aligned);
+
+REGISTER_DEVICE_IMPL(bezier_align_forward_impl, CPU, bezier_align_forward_cpu);
```
Comment (grimoire) [inline mmcv/ops/csrc/pytorch/cpu/bezier_align.cpp:566]: `bezier_align_forward_cpu` and `bezier_align_backward_cpu` are note necessary, register with `BezierAlignForwardCPULauncher` and `BezierAlignBackwardCPULauncher`

File: mmcv/ops/csrc/common/cuda/bezier_align_cuda_kernel.cuh [hunk 5]
```diff
@@ -0,0 +1,266 @@
+// Copyright (c) OpenMMLab. All rights reserved
+#ifndef BEZIER_ALIGN_CUDA_KERNEL_CUH
```
Comment (grimoire) [inline mmcv/ops/csrc/common/cuda/bezier_align_cuda_kernel.cuh:2]: As the code is modified from other repo. We should add some comments here, for example:
```dotnetcli
// modified from
// https://the.link.of.the.code.in.origin.repo
```

File: mmcv/ops/csrc/pytorch/cpu/bezier_align.cpp [hunk 6]
```diff
@@ -0,0 +1,567 @@
...
+  // reference in forward
+  // T v1 = input[y_low * width + x_low];
+  // T v2 = input[y_low * width + x_high];
+  // T v3 = input[y_high * width + x_low];
+  // T v4 = input[y_high * width + x_high];
+  // T val = (w1 * v1 + w2 * v2 + w3 * v3 + w4 * v4);
+
+  w1 = hy * hx, w2 = hy * lx, w3 = ly * hx, w4 = ly * lx;
+
+  return;
```
Comment (grimoire) [inline mmcv/ops/csrc/pytorch/cpu/bezier_align.cpp:351]: `return;` is not necessary.

File: mmcv/ops/csrc/pytorch/cpu/bezier_align.cpp [hunk 7]
```diff
@@ -0,0 +1,567 @@
...
+    T p6_x = offset_rois[11] * spatial_scale;
+    T p6_y = offset_rois[12] * spatial_scale;
+    T p7_x = offset_rois[9] * spatial_scale;
+    T p7_y = offset_rois[10] * spatial_scale;
+
+    T roi_width = std::max(std::abs(p0_x - p3_x), std::abs(p4_x - p7_x));
+    T roi_height = std::max(std::abs(p0_y - p3_y), std::abs(p4_y - p7_y));
+    if (aligned)
+    {
+      AT_ASSERTM(
```
Comment (grimoire) [inline mmcv/ops/csrc/pytorch/cpu/bezier_align.cpp:206]: The behavior is different from the cuda implementation.

File: tests/test_ops/test_bezier_align.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+                                               0.0625]]]])
+
+
+@pytest.mark.parametrize('device', [
+    'cpu',
+    pytest.param(
+        'cuda',
+        marks=pytest.mark.skipif(
+            not IS_CUDA_AVAILABLE, reason='requires CUDA support')),
+    pytest.param(
```
Comment (grimoire) [inline tests/test_ops/test_bezier_align.py:29]: As this op has not been supported by mlu yet. Please remove the mlu param here.

**Review by grimoire — APPROVED** [review]
LGTM

File: tests/test_ops/test_bezier_align.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+])
+@pytest.mark.parametrize('dtype', [
+    torch.float,
+    pytest.param(
+        torch.double,
+        marks=pytest.mark.skipif(
+            IS_MLU_AVAILABLE,
+            reason='MLU does not support for 64-bit floating point')),
+    torch.half
+])
```
Comment (zhouzaida) [inline tests/test_ops/test_bezier_align.py:38]: MLU does not support it now.

**Pull request merged**